### PR TITLE
[#182, Part 1] Refactor drawZoomRectangle to be a generic.

### PR DESCRIPTION
**Pull request by adamcw**
Implements #182.

Part 1: Refactors the drawZoomRectangle function so that it can be recycled to draw a selection rectangle for the fit functions as well.

### CODE REVIEW

File: client-js/app/elements/labrad-plot.ts [hunk 1]
```diff
@@ -1126,33 +1127,36 @@ export class Plot extends polymer.Base {
...
+      clip(y - this.margin.top, 0, this.height + offset)
+    ];
+  }
+
+
+  /**
    * Zoom into a selected rectangular region on the graph.
    */
-  private drawZoomRectangle(): void {
+  private drawRectangle(callback: Function): void {
```
Comment (maffoo) [inline client-js/app/elements/labrad-plot.ts]: As an alternate to taking a callback function, you can have this return promise, something like. `Promise<{rect, x1, y1, x2, y2}>`. Then you can use this with async/await for inline code. And you can still call it in callback style if desired by calling `then` on the returned promise.

**Comment by maffoo** [comment]
Consider using Promise instead of callbacks. Otherwise LGTM.

**Comment by adamcw** [comment]
@maffoo I made the function a bit more generic at the same time, so probably needs another look before I merge it.

File: client-js/app/elements/labrad-plot.ts [hunk 2]
```diff
@@ -1144,65 +1145,68 @@ export class Plot extends polymer.Base {
...
+          rect.style.display = 'block';
+        })
+        .on('mouseup', () => {
+          const [x, y] = this.mousePositionClipped(0);
+          d3.select(window)
+            .on('mousemove', null)
+            .on('mouseup', null);
+
+          if (x === originX || y === originY) {
+            return;
```
Comment (maffoo) [inline client-js/app/elements/labrad-plot.ts:1447]: I this case should probably reject the promise.

**Comment by maffoo** [comment]
One question about rejecting the promise if the box is degenerate (we might also want to allow this by pressing ESC mid-drag or something like that, though needn't be in this PR). Otherwise LGTM.

**Pull request merged**